File: Templates/union_find.py
class UnionFind:

    def __init__(self, size: int):
        self.root = list(range(size))
        self.rank = [0] * size

    def find(self, u: int) -> int:
        if self.root[u] != u:
            self.root[u] = self.find(self.root[u])
        return self.root[u]

    def union(self, u: int, v: int) -> None:
        root_u, root_v = self.find(u), self.find(v)

        if root_u == root_v:
            return False # cycle detected

        if self.rank[root_u] > self.rank[root_v]:
            self.root[root_v] = root_u
        elif self.rank[root_u] < self.rank[root_v]:
            self.root[root_u] = root_v
        else:
            self.root[root_u] = root_v
            self.rank[root_v] += 1

        return True

File: Templates/union_find.cpp
#include <vector>

class UnionFind {
public:
    explicit UnionFind(int size) : root(size), rank(size, 0) {
        for (int i = 0; i < size; ++i) {
            root[i] = i;
        }
    }

    int find(int u) {
        if (root[u] != u) {
            root[u] = find(root[u]);
        }
        return root[u];
    }

    bool union_set(int u, int v) {
        int root_u = find(u);
        int root_v = find(v);
        if (root_u == root_v) {
            return false; // cycle detected
        }
        if (rank[root_u] > rank[root_v]) {
            root[root_v] = root_u;
        } else if (rank[root_u] < rank[root_v]) {
            root[root_u] = root_v;
        } else {
            root[root_u] = root_v;
            rank[root_v] += 1;
        }
        return true;
    }

private:
    std::vector<int> root;
    std::vector<int> rank;
};
